kv/kvclient/kvcoord: TestTxnCoordSenderRetries failed

kv/kvclient/kvcoord.TestTxnCoordSenderRetries [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11110449?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11110449?buildTab=artifacts#/) on master @ [f295bd861a3a427652b19c2254d2401ebb4a3c8e](https://github.com/cockroachdb/cockroach/commits/f295bd861a3a427652b19c2254d2401ebb4a3c8e):


```
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestTxnCoordSenderRetries1742309064
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:98: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
=== CONT  TestTxnCoordSenderRetries
    dist_sender_server_test.go:3941: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestTxnCoordSenderRetries1742309064
--- FAIL: TestTxnCoordSenderRetries (21.81s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit
    --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit (0.07s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit/ReadCommitted
    dist_sender_server_test.go:3930: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3930
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3936
        	Error:      	Not equal: 
        	            	expected: false
        	            	actual  : true
        	Test:       	TestTxnCoordSenderRetries/write_too_old_in_staging_commit/ReadCommitted
        	Messages:   	TxnMetrics.ParallelCommitAutoRetries
        --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit/ReadCommitted (0.03s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Serializable
    dist_sender_server_test.go:3930: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3930
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3936
        	Error:      	Not equal: 
        	            	expected: false
        	            	actual  : true
        	Test:       	TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Serializable
        	Messages:   	TxnMetrics.ParallelCommitAutoRetries
        --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Serializable (0.02s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Snapshot
    dist_sender_server_test.go:3930: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3930
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3936
        	Error:      	Not equal: 
        	            	expected: false
        	            	actual  : true
        	Test:       	TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Snapshot
        	Messages:   	TxnMetrics.ParallelCommitAutoRetries
        --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Snapshot (0.02s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnCoordSenderRetries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30213

kv/kvclient/kvcoord.TestTxnCoordSenderRetries [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11136469?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11136469?buildTab=artifacts#/) on master @ [fcfa15d9694c0b840ac0c5c1b1489fa90a62efd8](https://github.com/cockroachdb/cockroach/commits/fcfa15d9694c0b840ac0c5c1b1489fa90a62efd8):


```
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestTxnCoordSenderRetries1466365788
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:98: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
=== CONT  TestTxnCoordSenderRetries
    dist_sender_server_test.go:3941: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestTxnCoordSenderRetries1466365788
--- FAIL: TestTxnCoordSenderRetries (22.02s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit
    --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit (0.07s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit/ReadCommitted
    dist_sender_server_test.go:3930: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3930
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3936
        	Error:      	Not equal: 
        	            	expected: false
        	            	actual  : true
        	Test:       	TestTxnCoordSenderRetries/write_too_old_in_staging_commit/ReadCommitted
        	Messages:   	TxnMetrics.ParallelCommitAutoRetries
        --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit/ReadCommitted (0.03s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Serializable
    dist_sender_server_test.go:3930: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3930
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3936
        	Error:      	Not equal: 
        	            	expected: false
        	            	actual  : true
        	Test:       	TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Serializable
        	Messages:   	TxnMetrics.ParallelCommitAutoRetries
        --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Serializable (0.02s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Snapshot
    dist_sender_server_test.go:3930: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3930
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3936
        	Error:      	Not equal: 
        	            	expected: false
        	            	actual  : true
        	Test:       	TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Snapshot
        	Messages:   	TxnMetrics.ParallelCommitAutoRetries
        --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Snapshot (0.03s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnCoordSenderRetries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvcoord.TestTxnCoordSenderRetries [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11172021?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11172021?buildTab=artifacts#/) on master @ [36665f664a0661b0d50649d16b1feb8e88afa56b](https://github.com/cockroachdb/cockroach/commits/36665f664a0661b0d50649d16b1feb8e88afa56b):


```
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestTxnCoordSenderRetries1051634340
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:98: 
        Test server was configured to route SQL queries to a secondary tenant (virtual cluster).
        If you are only seeing a test failure when this message appears, there may be a problem
        specific to cluster virtualization or multi-tenancy.
        
        To investigate, consider using "COCKROACH_TEST_TENANT=true" to force-enable just
        the secondary tenant in all runs (or, alternatively, "false" to force-disable), or use
        "COCKROACH_INTERNAL_DISABLE_METAMORPHIC_TESTING=false" to disable all random test variables altogether.
*
* WARNING: test tenant requested by configuration, but code organization prevents start!
* OSS binaries do not include enterprise features
*
=== CONT  TestTxnCoordSenderRetries
    dist_sender_server_test.go:3943: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestTxnCoordSenderRetries1051634340
--- FAIL: TestTxnCoordSenderRetries (21.69s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit
    --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit (0.07s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit/ReadCommitted
    dist_sender_server_test.go:3932: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3932
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3938
        	Error:      	Not equal: 
        	            	expected: false
        	            	actual  : true
        	Test:       	TestTxnCoordSenderRetries/write_too_old_in_staging_commit/ReadCommitted
        	Messages:   	TxnMetrics.ParallelCommitAutoRetries
        --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit/ReadCommitted (0.03s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Serializable
    dist_sender_server_test.go:3932: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3932
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3938
        	Error:      	Not equal: 
        	            	expected: false
        	            	actual  : true
        	Test:       	TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Serializable
        	Messages:   	TxnMetrics.ParallelCommitAutoRetries
        --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Serializable (0.02s)
=== RUN   TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Snapshot
    dist_sender_server_test.go:3932: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3932
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go:3938
        	Error:      	Not equal: 
        	            	expected: false
        	            	actual  : true
        	Test:       	TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Snapshot
        	Messages:   	TxnMetrics.ParallelCommitAutoRetries
        --- FAIL: TestTxnCoordSenderRetries/write_too_old_in_staging_commit/Snapshot (0.03s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnCoordSenderRetries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go
@@ -2012,6 +2012,11 @@ func TestTxnCoordSenderRetries(t *testing.T) {
 			return nil
 		}
 
+	// Disable load-based splitting to avoid unexpected range splits. The test
+	// operates between keys "a" and "z" and expects a single split point at "b".
+	// See the call to setupMultipleRanges below.
+	storeKnobs.DisableLoadBasedSplitting = true
+
 	var refreshSpansCondenseFilter atomic.Value
 	s := serverutils.StartServerOnly(t,
 		base.TestServerArgs{Knobs: base.TestingKnobs{
@@ -3181,10 +3186,10 @@ func TestTxnCoordSenderRetries(t *testing.T) {
 			// This test is like the previous one in that the commit batch succeeds at
 			// an updated timestamp, but this time the EndTxn puts the
 			// transaction in the STAGING state instead of COMMITTED because there had
-			// been previous write in a different batch. Like above, the commit is
+			// been a previous write in a different batch. Like above, the commit is
 			// successful since there are no refresh spans (the request will succeed
 			// after a server-side refresh).
-			name: "write too old in staging commit",
+			name: "write too old with put in staging commit",
 			beforeTxnStart: func(ctx context.Context, db *kv.DB) error {
 				return db.Put(ctx, "a", "orig")
 			},

